### ISSUE TITLE
[BUG] Named function references - arguments ignored

### ISSUE THREAD

**Issue by joewiz** [eXist-db/exist#3695]
**Describe the bug**

eXist ignores arguments passed to [named function references](https://www.w3.org/TR/xquery-31/#id-named-function-ref) and returns the function's identity instead of the results of the function applied to the arguments. For example:

```xquery
fn:string-length#1("foo")
```

... returns `fn:string-length#1` instead of `3`. 

BaseX and Saxon return the expected result, `3`.

**Expected behavior**

eXist should apply arguments to named function references instead of ignoring them and returning the named function's identity.

I think the relevant code is in https://github.com/eXist-db/exist/blob/develop/exist-core/src/main/java/org/exist/xquery/NamedFunctionReference.java.

**To reproduce**

Besides the xqsuite below, XQTS tests named function references extensively: https://github.com/w3c/qt3tests/blob/master/prod/NamedFunctionRef.xml (as [spotted by @dizzzz](https://github.com/eXist-db/exist/pull/3691#event-4174729995)). I suspect a fix if the tests below would pass, we'd pass this segment of XQTS too. 

```xquery
xquery version "3.1";

module namespace t="http://exist-db.org/xquery/test";

declare namespace test="http://exist-db.org/xquery/xqsuite";

declare
    %test:assertEquals(3)
function t:named-function-reference-direct() {
    string-length#1("foo")
};

declare
    %test:assertEquals(3)
function t:named-function-reference-wrapped() {
    (string-length#1)("foo")
};

declare
    %test:assertEquals(3)
function t:named-function-reference-apply() {
    apply(string-length#1, ["foo"])
};
```

This test returns:

```xml
<testsuite package="http://exist-db.org/xquery/test" timestamp="2021-01-06T09:39:52.997-05:00"
    tests="3" failures="0" errors="1" pending="0" time="PT0.006S">
    <testcase name="named-function-reference-apply" class="t:named-function-reference-apply"/>
    <testcase name="named-function-reference-direct" class="t:named-function-reference-direct">
        <error type="err:FOTY0013"
            message="The argument to fn:data() contains a function item. A function item other than an array cannot be atomized"
        />
    </testcase>
    <testcase name="named-function-reference-wrapped" class="t:named-function-reference-wrapped"/>
</testsuite>
```

Not included in the xqsuite is this expression:

```xquery
deep-equal(string-length#1("foo"), "3")
```

... which raises an unexpected error:

> Cannot compile xquery: err:XPST0017 Unexpectedly received 1 parameter(s) in call to function 'deep-equal()'. 

When this is added to the test suite as follows, the test won't even run:

```xquery
declare
    %test:assertTrue
function t:named-function-reference-deep-equal() {
    deep-equal(string-length#1("foo"), "3")
};
```

**Context (please always complete the following information):**
 - OS: macOS 11.1
 - eXist-db version: eXist 5.3.0-SNAPSHOT 35a76f25f0431220de95ff43752232bbf6a59efc 20210104165718
 - Java Version: 1.8.0_275

**Additional context**
- How is eXist-db installed? built from source
- Any custom changes in e.g. `conf.xml`? none

**Comment by dizzzz** [comment]
I saw similar issues in the XQTS suite. If I remember it was like `pi#0()`

**Comment by joewiz** [comment]
@dizzzz Indeed, your note in https://github.com/eXist-db/exist/pull/3691#event-4174729995 was what made me start looking into this. I chose `string-length` rather than `pi` for this report since it's a bit more straightforward. 

By the way, I've added a couple of tests to the report - `t:named-function-reference-variable` and `t:named-function-reference-arrow`.

**Comment by line-o** [comment]
Yeah, more -special- ways to call functions :)
This is definitely a parser issue from what I have learned making changes to it.
Our grammar simply does not include the case function reference called immediately.

**Comment by line-o** [comment]
This issue is fixed :tada: tested in 6.4.0
